lib/error_tracker/plugins/pruner.ex

```diff
@@ -0,0 +1,130 @@
+defmodule ErrorTracker.Plugins.Pruner do
```

Comment:
I don't know if this is the right approach. Let me explain:

I agree than removing resolving errors is a valid use case, but it can also be keeping the amount of occurrences under control, or just delete old occurrences which are not relevant anymore.

However, I'm not usre if we can or should encapsulate all those cases on the same plugin or it should be different ones.

I can think of using two plugins may be safer:

## `ResolvedPruner`

Allows to set the way to prune resolved errors. When an resolved error is deleted, all occurrences are dropped.

It would be this module code.

## `OccurrencesPruner`

It would remove occurrences (and the error if no more occurrences are stored).

It can use this options:
- `max_occurrences`: if the error has more than those # of occurrences, delete the old exceeding ones.
- `max_age`: delete the occurrences older than that amount of time
- `limit`: the amount of occurrences to delete per execution.
- `interval`: the interval in milliseconds between pruning runs.

---

That way we can support all cases I think, as keeping the DB size under control is not strictly the same as removing resolved errors.

It could be in an ideal world, but not all teams can resolve bugs as fast as they would like, and there are apps in the wild that may generate _tons_ of errors per minute depending on where is the bug.

Reply:
I have to think about this. Will get back to you.

Reply:
On the one hand I agree with this. Depending on the number of occurrences, there may be problems when removing high number of errors with high number of occurrences.

On the other hand having two different pruners feels too much to me. I would like to keep it as simple as possible in line with the plug&play philosophy.

Instead, I like your other proposal more: deleting the occurrences manually from the application code instead of relying on delete cascade. This would avoid situations when deleting a number of errors implies the deletion of an unbounded number of occurrences at once.

---

Getting deeper into this I am starting to think that we should limit the number of occurrences that are stored for each error. Aside from the considerations that we are discussing I do not think that it is very useful to store thousands or tens of thousands occurrences for an error. 

Maybe we should have a default (but configurable) limit for occurrences so we can be sure that errors won't have more occurrences than that. This would allow us to keep the deletion logic simple.

If someone decides to increase the default limits and experiences performance errors during pruning, they can always resort to manual pruning and tune the default options for their particular need.

What do you think?

Reply:
I've just pushed another commit that batches occurrence pruning before pruning the errors themselves. To avoid exposing too much configuration I've set the occurrence batch size to 1000 which should be reasonable enough (although I had initially set it to 10_000).

I still agree that keeping the number of occurrences under control could be desirable, but I think that is independent from pruning and we could handle it separately.